### PR TITLE
Tools

### PR DESCRIPTION
**Pull request by KennethNielsen**
This PR introduces parts of my standard python development tool chain into ixdat for evaluation. The tools falls into two different categories, "tools" and something which I generally call "tool runners".

Of tools this introduces only `pytest`as a test runner and `flake8` as a linter (takes care of some coding errors and coding style related stuff). flake8 is more or less standard if you want a fast and not too opinionated linter and I have over time come to prefer that over something like pylint. I did not include `black` to the configuration just yet, as I wanted to see if you are happy with it first.

**Tool runners** is likely the more interesting category, because those types of tools may not be too well known for scientists. These are tools to run your tools and perform tasks for you. You may ask, "why would I want a tool to run a tool?". The answer is that it can help you remember worthwhile options and in general make sure that it is run in an consistent manner and that tasks are performed easier and consistently.

As tools runners I have introduced `invoke` (http://www.pyinvoke.org/). The purpose of invoke is to easily make it possible and easy to do all the things/perform all the tasks, you need during development. Whenever you need to do something, you should check if invoke has a task for it and ask it to do it. If you perform a task that invoke hasn't been configured for and you do it more than twice, you should probably just go ahead and define an invoke task for it. In the olden days Unix developers would often use something like makefiles for this (https://en.wikipedia.org/wiki/Makefile), but they don't work on Windows which is the reason something like sphinx has both a makefile and a make bat file to compile the docs. Invoke aims to bridge that gap by providing as much of a platform independent command execution environment available as possible and by making its configuration file a pure python file, which means that you can yourself program your way around the whatever platform difference issues you may run into.

During development invoke commands to commonly use could be something like:

```bash
invoke lint
```
to run the linter

```bash
invoke tests
```
to run the tests

```bash
invoke checks
```
to run all the standard code checks (lint and test for now)

```bash
invoke clean
```
to delete all the temporary and cache files

```bash
invoke build-docs
```
to build the docs (not yet implemented)

to see a list of all available tasks do:
```bash
invoke --list
```

The second tool runner is `tox` (https://tox.readthedocs.io/en/latest/). tox is a tool to run all your QA tools in as consistent a manner as possible across as many of the supported python version as possible. It runs the QA tools in a virtual environment on a freshly installed version of your code. Because tox is meant mainly for continuous integration environments, it will try and perform all the tests on all the supported python versions. If they are unavailable, it will report an error, for this purpose I have configured an invoke task to run tox only on the python version that are available.

To run tox purely do:
```bash
tox
```

to run in parallel do:
```bash
tox -p auto
```

To run tox only on available python versions do
```bash
invoke tox
```

I tried to gather all the information that developers will needs about the tools in the TOOLS.rst file. Please give that a look.

NOTE: This is WIP since I know it has a directory conflict with #2, which I will resolve here when #2 is merged.

EDIT: I completely forgot to mention that the PR also includes a git hook, which is a mechanism to run tests and checks before pushing commits to a remote branch. Please see the details of it and installation instructions in TOOLS.rst.

### CODE REVIEW

File: TOOLS.rst
Comment (ScottSoren) [inline TOOLS.rst:18]: "Quality Asssureance (**QA**)", for us damn academics who don't know what QA is :wink:

**Review by ScottSoren — CHANGES_REQUESTED** [review]
These tools look great! 
I'll admit that I haven't got it running yet, but may as well get these comments to you anyway before I disappear for meetings the rest of the afternoon.
Most of the comments are questions which are more for my understanding than anything else but may as well get answers in the documentation.

My main concern about it is whether it adds a barrier of complexity that deters experimentalists who are not software developers from contributing. That is not a critique of the tools, but just a note that I think we have to make ixdat (or at least parts of ixdat) develop-able without the tools. It's more a governance thing, who's responsibility is it to do QA and when, and a luxury problem for a hypothesized time in the future when we have lots of contributers.

File: TOOLS.rst
Comment (ScottSoren) [inline TOOLS.rst:132]: I'm using anaconda and get the following, in the PyCharm Ipython console:
```
>>> import sys
>>> sys.executable
'C:\\ProgramData\\Anaconda3\\envs\\ixdat\\python.exe'
```
so...
/c/ProgramData/Anaconda3/envs/ixdat/python.exe
?

File: TOOLS.rst
Comment (ScottSoren) [inline TOOLS.rst:28]: I'm missing some instructions, with the level of detail of the git hooks instructions, on how to get invoke working.
Naively checking out the [tools] branch, navigating the anaconda powershell prompt to the project folder, and typing 
`> invoke tox` 
gives the error
`invoke : The term 'invoke' is not recognized...`
I'm not completely surprised that something is missing, but I don't know what it is.

Comment (ScottSoren) [reply TOOLS.rst:28]: Ah, yeah, of course:
```
> pip install invoke
> pip install tox
```
Works in anaconda prompt now but not git bash.

File: TOOLS.rst
Comment (ScottSoren) [reply TOOLS.rst:132]: It's not working for me. After adding the line to ~ /.bashrc and sourcing ~/.bashrc, it's there:
```
$ echo $PATH
/c/ProgramData/Anaconda3/envs/ixdat: ....
```
but `invoke` remains unrecognized by git.
I did remember to install invoke in ixdat's python environment in PyCharm (pip install invoke in Anaconda only gets it onto its `(base)` environment).
So I think some things are missing in these instructions.

**Review by ScottSoren — CHANGES_REQUESTED** [review]
Okay, I tried to follow the instructions in tools.rst to get up and running, and find they are not complete, at least not for Windows with Anaconda Python and PyCharm.
There are comments with details.
I will keep playing around and update here if I get it working.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:59]: Fixed the formulations

I'm not quite sure where you meant to add the sentence. I changed it to:
```
* **setup.cfg** (not to be confused with setup.py, which is used for
  building the package) If the tools support defining settings in the
  shared configuration file ``setup.cfg``, then they belong there
  no-matter how the tools are run. If not (settings may only be
  available at the command line), see the next points.
```

Was that what you had in mind?

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:63]: Maybe I didn't formulate it clearly. If the tool support being configured in setup.cfg, then it is configured there, always. setup.cfg is an attempt to clean up the mess of ``.something_rc`` files that used to live in the root of projects, because they all had their own configuration files. Unfortunately, not all tools support it, but quite many does, so we gather those.

If, and only if, the tool doesn't support configuration via setup.cfg (or its own configuration file), but only allows configuration on the command line, then we move the actual settings into tox.ini or tasks.py. And of course the actual name of the command itself will always have to be listed in one of the these two files.

As for the **tox precedence**, that is unfortunately the only way that it can be, since tox' configuration is just an ini-file and it can't read command lines from tasks.py, but seeing as tasks.py is a real python-file, we can do it the other way. But keep in mind that the precedence that you mentioned is only ever relevant/present for **QA tools**, of which I have already configured 2 (test-runner, linter) out of the 4 (test runner, linter, code format checker and possibly static type checker) that one might usually have and I plan to add the 3 also (code format checker), so most of this shared configuration is already done.

Also keep in the mind that this is relevant **only for the developer who wants to add or tweak a tool**. It is totally invisible for developers who wants to just use it. An ordinary developer who wants to take the dive and start using the tools and tasks need know only one command **invoke** (just like make in the old days) and she will ask invoke for everything she want's to know about what she can do (or consult the last sections of "Getting started developing", which will also only mention invoke commands). That is in fact the whole point about invoke, to make it easier for new developers by making it so that they do not need to know the names and common arguments of all the tools.

**tox** becomes relevant only for "top" developers or maintainers, who might want to run the tests in all the pythons they have available, to check cross python version compatibility locally maybe before a release. And it is used by the CI platforms such as Github Actions (which was one of the large motivators for setting tox up in the first place). And it is used by the commit hooks purely because it makes it possible to run all the QA in parallel which makes it run quite a bit faster, which you really want if you are waiting to see your push finish.

Does this alleviate your concern?

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:90]: Absolutely. I will make a general comment about the concern of what becomes mandatory in this PR (hint hint, nothing does). A small note though, git doesn't distribute hooks I think out of a safety concern, but it works out fine for an opt-in policy as well.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:140]: Good point. I converted all the command line examples.

File: setup.cfg
Comment (KennethNielsen) [reply setup.cfg:2]: **THAT** is 100% and totally at the discretion of the project BDFL (https://en.wikipedia.org/wiki/Benevolent_dictator_for_life). Line length is something that people get religious about and I don't want to spend 2 seconds discussing it. I wrote 95 here, because in other projects we have been going with recommendations that have been mentioned in a few different places, which is aim for 90 but allow up to 95, which essentially means that the developer should break at 90 unless there is a good reason, like a long link or something like that, which can be finished in the last 5. But since linters aren't that intelligent, it needs to allow up to 95 if that is the policy. But I will put there whatever you say.

File: setup.py
Comment (KennethNielsen) [reply setup.py:7]: Yeah. So, I can see that there is a bit of a misunderstanding, no doubt because I didn't explain it properly. It isn't that tox uses a setup.py file for anything special. It uses it for the same thing that we always do, which is to build a package. Every single time tox runs, it will build a package and install it into its own virtual environments (it has one for each python version and task it performs). It does this to make sure that e.g. the test will be run exactly on the code that will be distributed (the package) and not the code which is in the directory. The distinction is subtle but really important.

In any case. Because tox actually uses it to build a package, then it is actually probably able to build a package. What **is**  missing are the trove classifiers (https://pypi.org/classifiers/), what **is probably** missing is read through by you of the meta data in ``__init__.py`` to make sure it is as you wish and what **may be** missing fairly soon is something on data-files, which you need to tell setup tools about separately.

I will add the comment at the top.

File: setup.py
Comment (KennethNielsen) [reply setup.py:23]: Yes. Updated.

File: setup.py
Comment (KennethNielsen) [reply setup.py:32]: Yeah sorry. I shamelessly stole these from a common packaging guide, and therefore didn't give them much concern. Will update the docstring and now that I think about it add a reference.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:87]: Yeah. There isn't actually a git hook for that, since building the package isn't related to git. But I see your point. Normally this is done by settings up a CI tool (it is the next thing that I will propose) so that the QA tools are run automatically on a server, on the HEAD of master and on the head of every single PR ever, so that you will know immediately whether the code passes QA. In that case, actually building a package will then happen on code you likely checked our and package without modification, and that you therefore know passes QA.

You could build a soft hook by making the build of the package a task in invoke, that will run tox first on the available environments and exist with error if they do not pass, before actually building the package. That is commonly regarded a sufficient check. What do you think?

File: setup.cfg
Comment (KennethNielsen) [reply setup.cfg:8]: Yes. I have added comments.

File: tools/hooks/pre-push
Comment (KennethNielsen) [reply tools/hooks/pre-push:21]: Yeah. Agreed. I've been thinking about this since I wrote it. So the thing is, you are completely right about the analysis, we do not need to run the tests on multiple python verisons. The only thing is that tox has the ability to run the checks (test, lint etc) in parallel, which is really great since it speeds things up when the code base grows even a little and you don't ever want these hooks to take time since then they will become an annoyance and people will switch them off.

I mean to fix this by making a special tox task in invoke, that runs the tests only in one python environment. I will update the code and you can have a look.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py:23]: A added to the docstring the explanation of why we load the command to tox and I added an example of the ``.ini`` syntax we are parsing. I hope this is enough. WRT to "complexity propogation and not complexity reduction" the goal is to not duplicate settings, which in my view is as bad as duplication code. I hope with the comments that I have alleviated the concerns.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py:35]: Yeah, apologize, I made quite a few bad assumptions about what is known about these tools. It the execution context argument automatically passed in by invoke, which gives access to a range of different commands in a operating system independent way (http://docs.pyinvoke.org/en/stable/api/context.html). I updated the docstrings.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py:99]: I don't know to be honest. The problem is that on my system the python3.6 command **is** the reduced version, so I don't really know what else to do. In any case, making sure that all the pythons are there is mainly the task of the virtual machines that will run the CI. If we can just make sure that someone who runs the invoke tox task doesn't get any false negatives I think this is alright. Unresolve if you disagree.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py:109]: Wishful thinking. It's all just Linux ... Thank, fixed.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py:135]: Fixed. ``c`` is a badly named version of the same ``context`` as before and this one has a dryrun to make it possible to see what it will delete without actually deleting it.

File: tools/hooks/pre-push
Comment (KennethNielsen) [reply tools/hooks/pre-push:1]: Yes

File: tox.ini
Comment (KennethNielsen) [reply tox.ini:1]: My bad. Autogenerated file, fixed it.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:18]: Ah dang, I though I defined it at first use :confused:  ... did for CI though though, maybe that's why I'm confused.

Fixed.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:28]: Yeah. I added them all to `requirements-dev.txt` so `pip install -r requirements-dev.txt` in the anaconda prompt should get you everything including pytest, flake8, invoke and tox

Comment (KennethNielsen) [reply TOOLS.rst:28]: I added the following installation instruction section at the top of the file:

```rst
Install instructions
--------------------

Install all development tools by executing::

  pip install -r requirements-dev.txt

while being at the git archive root in your virtual environment or
anaconda environment.
```

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:132]: Yeah. I'm slowly learning that while anaconda is really good for getting a lot of high quality, high speed scientific code into the hands of scientists in a hurry I find that it is actually somewhat harder to figure out exactly what goes on. I actually intended to update this later specifically for anaconda (since I just uninstalled it in favor of an ordinary python installation and virtual environments, but since you show an interest in the hooks I think I will go back to get this to work and properly documentet.

Comment (KennethNielsen) [reply TOOLS.rst]: Actually, can I persuade you do me one favor. Can I get you to run:

```cmd
echo %path%
```

in your anaconda prompt and write the results here.

**Comment by KennethNielsen** [comment]
Thanks for the review @ScottSoren 

Ok, so I resolved most of the issues with a few pending feedback from you @ScottSoren.  There are 2 things I wanted to comment in directly:

**1**
**You mentioned a concern about the development tools creating to high a barrier for entry for new developers.** I can assure you that I share that concern and I think it is important that **we maintain one low barrier path for contribution** (just because I like programming and sharing what I know, doesn't mean that I insist that everybody must learn;)). There is however not actually anything in this PR that conflicts with that. Using any of the regular tools or the tool runners are optional. As long as the git hook isn't enabled (which is an opt-in) there isn't anything that will interfere with the casual/scientific contributor.

What I suggest as the next step is to enable a continuous integration service, which will display a CI status (right besides the "merge-ability" status at the bottom of every PR, like so:
billede
(note, the status reduces to just a single line if all is green)

Now, I would suggest enabling the **display** of such status no-matter what, so that especially we, but also the interested new contributor, can see the CI status straight away (as a maintainer you **have** to reduce your manual chores as much as possible, trust me). However, whether an all green status is required for merge or not, is a decision. And we can set github up for the archive in a manner that reflects that decision. If you want to insist that all is green before merge, we can ask github to enforce that, but if you want to allow for merging "dirty" and then fixing the QA issues manually afterwards yourself (which is a policy I have seen some places), then we can just choose not to have github enforce it, in which case the CI status will be just for information.

**2**
**The sharing of settings/command line between tox and invoke for QA tools**. I wanted to mention here, that I'm willing to have that challenged. It is the first time I have done it like this, so I don't have prior experience to base the suggestion on. I have however been burned by not knowing with which settings a tool was run, which is the motivator for it, and in my view, avoiding multiple definitions of settings and commands is worth adding a little complexity. However, if you, even with the updates of comments I have now made, don't think it is worth it I will remove it. The duplication is limited at this point anyway.

File: tools/hooks/pre-push
Comment (KennethNielsen) [reply tools/hooks/pre-push]: I updated the code and gave the tox task a ``--single`` argument and I changed the hooks to use that. In my view this is the best of both worlds. We will have parrallel execution and independently of what invoke checks does, the pre-push hook will execute the same exact tests that the CI tools will on github, which means that you can't get in that annoying situation where you don't notice that something fails until you already made the PR and then you have to do a small fix-commit. If you want to test it, remember to update your hook if you copied the file rather than soft link it. I will mark the comment resolved because I think that this solves your comment, but re-open if you don't agree.

File: setup.py
Comment (KennethNielsen) [reply setup.py:11]: Comment added.

File: TOOLS.rst
Comment (ScottSoren) [reply TOOLS.rst:98]: I like that idea! `invoke build`?

File: tasks.py
Comment (ScottSoren) [reply tasks.py]: As you taught me, it's not dupliation of code which is evil, it is duplication of *information*. If we decide to use flake8 as the linting command in invoke, and flake8 as the linting command in tox, there's information in both of these. Another question is, why would someone chage the command? A possible reason is an updated linter, in which case we'd want the change in both tox and invoke. But another possibility is an update in tox OR invoke that make them preferable to use with another linting command. Then we might want to change it in one place and not the other. So there's information in the fact that we want to use flake8 in both tox and invoke.

I prefer duplicating the settings...

File: tasks.py
Comment (ScottSoren) [reply tasks.py:151]: got it.

File: TOOLS.rst
Comment (ScottSoren) [reply TOOLS.rst:74]: Thanks a lot for this. I am happy about the different roles of invoke and tox, and the relationship between the two.

I however, after thinkng more and understanding it better, prefer that they don't share settings, due to the complexity of the implementation (and its explanation in this .rst), and my argument here: https://github.com/ixdat/ixdat/pull/3#discussion_r556428028

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:98]: Yeah :)

I mean, the whole process consist of 3 steps maybe, with sub-steps. It could be something like (just brainstorming here):

Release:
 * Check that there are release notes in the documentation, if we use such
 * Update the version number and commit
 * Tag the release in git (using "v....." tags in git is common practice)
 * Push both the new commit and the tag

Build:
 * Perform local QA
 * Build
 
Upload:
 * Upload to PyPI

In principle any one of these could be automated and you could break this up in 3, 2 or 1 (depending on temper) tasks in invoke and have the whole release process scripted. Of course there are things which cannot be scripted like actually writing the release notes and making a new milestone on github of we use those, so we should still have a release procedure to follow, but it can be made significantly shorter and more re-producable.

Let's look into that separately.

PS: If you are interested you can read the SoCo release procedure here. It is missing some thing compared to above, because we used to have our CI tool (TravisCI) build and publish the packages, but that service is no longer available, at least not from there.

File: tasks.py
Comment (KennethNielsen) [reply tasks.py]: Agreed about the first part (duplication of information), not necessarily about the latter (that there is unique information there), but it's not important enough to use more time on :) I will make the change in a second.

Comment (KennethNielsen) [reply tasks.py]: I pushed the change.

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:74]: Done, as per a comment above. Of course I forgot to update the TOOLS.rst to reflect the change. I will force push a amended commit in a second.

File: TOOLS.rst
Comment (ScottSoren) [reply TOOLS.rst]: image

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst:74]: With that change the section in TOOLS.rst becomes:

File: TOOLS.rst
Comment (KennethNielsen) [reply TOOLS.rst]: Great! :|

I will see about re-installing anaconda shortly, but it will delay this PR a little further. I would prefer actually to merge it and fix it later as time allows (since it is marked with a TODO in TOOLS.rst) but I will leave that up to you.

Comment (KennethNielsen) [reply TOOLS.rst]: That's fun. That command is mentioned like ... everywhere. Grumble grumble.

File: TOOLS.rst
Comment (ScottSoren) [reply TOOLS.rst]: Looks great with the formating, and the explanation!

File: setup.py
Comment (ScottSoren) [reply setup.py:11]: Thanks! The file looks great now.

File: tasks.py
Comment (ScottSoren) [reply tasks.py]: Looks good. 
Wheth or not a little bit of information is now duplicated I think it lowers the barrier a bit to understanding the tools, and I think that that's worth it. Thanks!
The docstrings on this module also look great now.

However, I think it still looks in tox.ini for invoke settings. No effective difference so long as they are the same, but contradicts the .rst. Can update later - I can do so while getting the tools running on my machine in a new branch.

File: tasks.py
Comment (ScottSoren) [reply tasks.py:27]: Thanks, updated docstring looks great!

**Comment by ScottSoren** [comment]
Looks great! 
Thanks for the additional documentation and in one case simplification. I want to re-iterate that I think these tools are fantastic, will definitely use them, and want it to be as easy as possible for new developers to use them as they get more into ixdat. I think that now they can find all the explanations they need in the code to enjoy that process. 
I do not yet have the tools running in my PyCharm + Anaconda environment, due potentially to its quirks. I will make a new branch to do so. I also think we should merge this now! 
Pending @kkrempl's read-through and approval.

**Review by ScottSoren — APPROVED** [review]

File: tasks.py
Comment (ScottSoren) [reply tasks.py]: Never mind, I was looking at it one commit behind. It looks good!

**Review by kkrempl — APPROVED** [review]

**Comment by ScottSoren** [comment]
Merged these tools! Thanks @KennethNielsen.
Some additions expected soon to tools.rst to cover e.g. *Anaconda+PyCharm+git bash*

**Pull request merged**
